### ISSUE TITLE
The correct way to theme Uniform components that wrap Vaadin

### ISSUE THREAD

**Issue by warpech** [Starcounter/uniform.css#95]
Right now, `uni-date-picker` is themed by importing `vaadin-material-theme/vaadin-date-picker.html` before importing `/uniform.css/components/uni-date-picker/uni-date-picker.html`.

This is perhaps not good, because `/uniform.css/components/uni-date-picker/uni-date-picker.html` imports 
 `vaadin-date-picker/vaadin-date-picker.html`, and that file already includes another theme (Lumo).

According to @tomalec, the correct way is to import  `vaadin-date-picker/src/vaadin-date-picker.html`, which is themeless.

But then how do we make sure that there is always some default theme?

**Comment by tomalec** [comment]
We talked about it briefly with @platosha 
The problem seems not to have easy solution. As `vaadin-date-picker/src/vaadin-date-picker.html` have absolutely no look, and is not usable without a theme.

Also you cannot load two different themes, as they will be used together and collide.

We need to find a way to overcome this somehow, as above means there are no themes. There is just one element definition (with one style) or the other definition (with another styling).

**Comment by platosha** [comment]
Users might want add a small customisation on top of the existing theme. Applying all the found theme modules in `Vaadin.ThemableMixin` simplifies this kind of theme extensions. We thought that customising an existing theme is a more common use case, than creating and applying an entirely new theme.

> But then how do we make sure that there is always some default theme?

Starting with the recent major (`<vaadin-date-picker>` v3):
- Themes are shipped in the components. Users are expected to use `theme/lumo/vaadin-date-picker.html` for the Lumo themed component.
- If there would be another theme shipped, it would come as a separate import, e. g., `theme/material/vaadin-date-picker.html` for the Material one.
- The root import `vaadin-date-picker.html` is an alias for the default theme, `theme/lumo/vaadin-date-picker.html`, this is how users get some theme by default.
- Using external themes is possible, but, as mentioned, users have to use `src/vaadin-date-picker.html` with an external theme.

In the previous major versions of the components (`<vaadin-date-picker> v2 had one), we had a default theme looking kind of like a browser default, which magically disappears when there is at least one external theme module to prevent any collisions. We decided to drop that one, here is why:
- It’s unthemed look-and-feel was better, but still not good to compete with the theme.
- With the theme, the default theme turns off, but remains delivered in the component’s HTML to the browser, despite it was totally unnecessary.
- It is confusing to theme authors, that once you define a new empty theme for a component, the component look-and-feel changes quite a bit because the default theme turns off.

At this point, I would recommend creating a bare and simple uni-date-picker theme with the look-and-feel that matches the rest of the uniform.css, and using that with an `src/vaadin-date-picker.html` import.

**Comment by tomalec** [comment]
Basically, it means that is a Vaadin element is used by app dev it's not themeable any longer and we are locked with the theme he/she picked. When another app dev picks the other theme, both are broken.

The solutions I see:

1. > At this point, I would recommend creating a bare and simple uni-date-picker theme with the look-and-feel that matches the rest of the uniform.css, and using that with an `src/vaadin-date-picker.html` import.
   
   That requires us to create `uni-` wrappers for all Vaadin elements, and kindly ask app devs not to use `Vaadin-elements` - technically no way to force that, probably not so desired.
2. Hack, monkeypatch Vaadin's `ThemableMixin` to use only the first theme for given element.

   Hacky, disables the feature of vaadin themes to enhance theme without forking a theme. I haven't seen any Starcounter user using that. Enhancing individual instance is doable in shadow DOM compositions, scenario is doable via fork (as it is global anyway). We could also make it so that enhancements to existing theme reuires additional attribute, like `<dom-module id="smth-for-material-date-picker" theme-for="vaadin-date-picker" enhancement>`.

   It will still load both themes, even those disregarded (traffic cost).

3. Ask devs to import vaadin's element source separately from the theme:
   Which will be hard to control and make sure nobody loads default Lumo, that will break everything.
   Then,

   1. And set the theme arbitrary in `app-shell.html`/`uniform.css`/Blending app (not sure how)
   2. Let developer load the default - Lumo - theme, using Starcounter- specific element `<theme-link>` which would give us some control, but this also have downsides:
      - not a `link` anymore so will break the `<head>` - probably not a big deal if we use it in shadow roots,
      - I still don't know how to cover cases of more complex dependencies, without processing the document and dependency graphs manually before import.

**Comment by warpech** [comment]
Thanks for the explanation, but I disagree with the recommendation in the last paragraph. We don't want to go through a hassle of creating our own look of the date picker, at least not right now.

Let's hardcode the use of Lumo theme in `<uni-date-picker>`.

95% of users will be fine with Lumo and its configurability via CSS custom properties. 

5% of users who hate Lumo, can replace the implementation of `<uni-date-picker>` to include Material theme or their own theme.

**Comment by tomalec** [comment]
On today's meeting with @malx122, @platosha, @warpech, @PatrykSzwer and me, we agreed to hold theming feature in Starcounter. Then maybe reconsider it once after CSS Shadow parts are shipped, or we have a customer demanding more than is available.

The reasons:
TL;DR Theeming in currently available shape seems to provide small value at big cost.

- [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it) @malx122 expressed a need for customizing the default theme of our Design System as a priority. That's something we should start with, make sure it completely works and see how far we can get with it.
- Vaadin themes are extremely problematic to apply in our case when all HTML imports are distributed and cannot be controlled by single authority. App devs from many companies and solution owners could load their elements, in hundreds of their views. If any of them loads a different theme, all the elements are broken. Protecting against that is a huge effort to develop and for runtime.
- Changing themes is technically almost the same as changing custom element definitions, but visual difference is not so astonishing.
- If solution owner wants to drastically change the look and feel of `uni-` or `any-` elements used in entire solution, he/she can currently do that with running a different or forked version of [https://github.com/Starcounter/StarcounterClientFiles](https://github.com/Starcounter/StarcounterClientFiles).
- Hopefully once we will have CSS Shadow Parts (`::theme, ::part`) we will be able to achive way much more, with a fraction of this effort.
